KEP: Support for CSI Plugins on Windows Nodes

ddebroy opened the pull request:
Proposal for enhancements to support CSI plugins on Windows nodes.

ddebroy commented:
/sig storage

ddebroy commented:
/assign @yujuhong @msau42 @PatrickLang @jingxu97 
PTAL

PatrickLang commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Non-Goals
+
+1. Support CSI controller plugin operations from Windows nodes: This may be considered in the future but not an immediate priority. If all the worker nodes in the cluster are Windows nodes and Linux master nodes have scheduling disabled then CSI controller plugins cannot be scheduled for now.
```
would be good to clarify that this doesn't necessarily require privilege and could already be done today.

PatrickLang commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+      volumes:
+        - name: csi-proxy-pipe
+          hostPath: 
+            path: \\.\pipe\csi-proxy
```
shouldn't this be `unix:/csi/csi.sock` ?

ddebroy replied:
This is the named pipe exposed by csi-proxy.exe to the CSI node plugins. In the path between CSI node plugins and csi-proxy.exe, csi-proxy.exe is the host based process that listens to RPC calls and services them. That scenario appears to always fail in Windows when using domain sockets with permission denied errors. It does work when a host/containerized process invokes RPCs on a process in a container which is the kubelet.exe/CSI node driver registrar => CSI node plugin.

PatrickLang commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+    // StageDisk partitions a disk device and formats the partition with NTFS (if the disk has not been partitioned or formatted)
+    // The resulting volume is mounted at the requested global staging path.
+    rpc StageDisk(StageDiskRequest) returns (StageDiskResponse) {}
```
Might need ResizePartition() too to support volume expansion

ddebroy replied:
Added.

PatrickLang commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, an admission webhook will need to be implemented and deployed that will reject all containers that bind mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have `privileged` flag set in the pod's securityCont ...
```
should this be unix socket or pipe?

PatrickLang replied:
https://github.com/open-policy-agent/gatekeeper would also work great - could write a policy such as only allow unix socket mounts if in kube-system namespace

wongma7 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskNumberWithLocation finds a disk device with requested Location <Bus, Target, LUN ID>
+    rpc GetDiskNumberWithLocation(GetDiskNumberWithLocationRequest) returns (GetDiskNumberWithLocationResponse) {}
```
What if the Location information isn't available anywhere? The AWS EBS driver would like to run a `ebsnvme-id.exe` on the host to map the cloud's reported volume-ID to the disk number: https://github.com/kubernetes/kubernetes/pull/79552/files#diff-238adf22aa109b4ad41399fe48280ec8R214. GCE PD may be similar, it runs some powershell cmdlets on the host: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/attacher.go#L147

Adding a proxy call to `Get-Disk` to the API will be a suitable workaround for EBS (ebsnvme-id.exe is not strictly necessary, at the moment `Get-Disk` is guaranteed to return the same information), but I'm not sure about GCE PD.

msau42 replied:
@jingxu97 to determine if this api works for GCE PD or we need a special service to map PD devices.

ddebroy replied:
As part of the prototype to make sure everything does work end-2-end from a OS plumbing perspective, I ported the GCE PD CSI driver to work on Windows node. As part of that, I got the SCSI Page 83 queries and mapping working (required for the mapping of GCE PD and potentially by other vendors too) by creating a generic version of the `Get-GcePdName` cmdlet that directly issues `IOCTL_STORAGE_QUERY_PROPERTY`. The generic wrapper for that is exposed through the `GetDiskNumberWithSCSIID` API mentioned in the KEP. 

For the `GetDiskNumberWithLocation` API, I was indeed planning to use the information returned through `Get-Disk` @wongma7.

pjh replied:
I smiled when I saw `GetDiskNumberWithSCSIID`, I knew exactly why it was needed...

The GCE Persistent Disk driver (in-tree at least, haven't looked at the CSI version but assuming it's the same...) needs to map the PD "friendly name" to a Windows device id. We're starting to investigate if we can provide this information natively through the `Get-Disk` cmdlet (https://github.com/kubernetes/kubernetes/issues/74674). Even if we can change GCE to support this though, it looks like the PD driver would still end up using `GetDiskNumberWithSCSIID` rather than `GetDiskNumberWithLocation` since the latter will use <Bus, Target, LUN ID> as its key. Is that right?

Is the PD driver the only storage driver that uses such a "friendly name" for identifying the volume the pod wants to use? If not then perhaps an API method that accepts that as the argument would be useful; if so then `GetDiskNumberWithSCSIID` seems probably suitable.

msau42 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+A CSI Node Plugin will interact with csi-proxy.exe using a named pipe: `\\.\pipe\csi-proxy` (exposed by csi-proxy.exe). The pipe will need to be mounted into the Node Plugin container from the host using the pod's volume mount specifications. Note that domain sockets cannot be used in this scenario since Windows blocks a containerized process from interacting with a host process that is listening ...
+
+A GRPC based interface will be used by CSI Node Plugins to invoke privileged operations through csi-proxy.exe. The GRPC API will be versioned and any release of csi-proxy.exe binary will maintain backward compatibility across all prior versions of the API. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins have been config ...
```
backwards compatibility even for alpha?

msau42 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+07/16/2019: Initial KEP drafted
+
+## Drawbacks
```
Another drawback is that the proxy only supports standard block and SMB volume types.  Storage systems with custom clients/protocols will need to install and talk to their own proxies.

Can the Daemonset for these kinds of CSI drivers create a hostpath and install their own proxies on the host directly?

ddebroy replied:
Good point - I will call this out.

yujuhong replied:
In that case, the proxy will use a different name for the pipe, right? E.g.,`\\.\pipe\csi-custom-proxy`

ddebroy replied:
Addressed at https://github.com/kubernetes/enhancements/pull/1141/files#r305145983 and also added an API Alternatives section outlining an approach to this.

msau42 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained.
+
+#### New Component: CSI Proxy 
```
To help visualize where this new component fits in, can you add a diagram similar to https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface_diagram1.png showing the communication channels between kubelet, sidecar and driver?

yujuhong replied:
Thanks for the link. This helps me visualize the components we have today much better.

ddebroy replied:
Embedded the above diagram in the previous section.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Implementation Details
+
+CSI Node Plugins listen on domain sockets and respond to CSI API requests sent over GRPC from a container orchestrator like Kubernetes. They are responsible all storage management operations scoped around a specific node that are typically necessary after a CSI Controller Plugin has finished provisioning a Persistent Volume and attached it to the node. In Kubernetes, the CSI Node API calls are in ...
```
For folks (like me) who are not familiar with the sockets used by CSI, it might be easier to digest if you list them here and state who creates and who use them and for what purposes.

ddebroy replied:
Does this diagram (already embedded in the KEP): https://ddebroywin1.s3-us-west-2.amazonaws.com/csi-proxy.png help?

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+#### Enhancements in Kubelet Plugin Watcher
+
+Registration of CSI Node Plugins on a Kubernetes node is handled by the Kubelet plugin watcher using the fsnotify package. This component needs to convert  paths detected by fsnotify to Windows paths in handleCreateEvent() and handleDeleteEvent() before the paths are passed to AddOrUpdatePlugin() RemovePlugin() routines in desiredStateOfTheWorld.
```
s/convert  paths/convert paths

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+#### Enhancements in Kubelet Plugin Watcher
+
+Registration of CSI Node Plugins on a Kubernetes node is handled by the Kubelet plugin watcher using the fsnotify package. This component needs to convert  paths detected by fsnotify to Windows paths in handleCreateEvent() and handleDeleteEvent() before the paths are passed to AddOrUpdatePlugin() RemovePlugin() routines in desiredStateOfTheWorld.
```
We should have a better library for converting Windows path in k8s...

ddebroy replied:
Agree - an universal Window "path normalizer" that is used across the board. Today there appears to be islands of code doing just that. Will mention it here.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+A CSI Node Plugin will interact with csi-proxy.exe using a named pipe: `\\.\pipe\csi-proxy` (exposed by csi-proxy.exe). The pipe will need to be mounted into the Node Plugin container from the host using the pod's volume mount specifications. Note that domain sockets cannot be used in this scenario since Windows blocks a containerized process from interacting with a host process that is listening ...
+
+A GRPC based interface will be used by CSI Node Plugins to invoke privileged operations through csi-proxy.exe. The GRPC API will be versioned and any release of csi-proxy.exe binary will maintain backward compatibility across all prior versions of the API. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins have been config ...
+
```
So for upgrade -- users will need to upgrade `csi-proxy.exe` first before upgrading the plugins, right?

The API version used by `csi-proxy.exe` should be equal or greater than the maximum API version used by all plugins.

ddebroy replied:
Yes - will clarify as you mentioned above.

ddebroy replied:
Clarified in the section around Upgrade/Downgrade strategy

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
````diff
@@ -0,0 +1,401 @@
+}
+```
+
+In accordance with standard Kubernetes conventions, the above API will be introduces as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. New RPCs and enhancements to parameters will be introduced through v1alpha1 and graduate to v1beta1 and v1 versions as they mature.
````
How would the the development and release workflow look like when CSI is updated and a new RPC call is needed in the proxy service?

ddebroy replied:
A good example of this is the new Resize API in the node plugin interface which requires a ResizeParition operation on the node plugin interface as well. In such cases, we will have to fast-track the API to v1 in the proxy. Given the proxy will be an out-of-tree component, such fast-track exceptions should be possible. I will detail this out here along with the need to anticipate potential changes in the proxy interface when CSI changes are considered.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+#### Enhancements in CSI Node Plugins
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above.
```
Will the same CSI plugin work for both Linux and Windows, or will separate codebase/images be maintained?

ddebroy replied:
Images will definitely be separate as the Windows plugin will be based on Windows base images and the Linux node plugin will be based on a Linux distro of choice. The codebase can be same. I will point this out specifically.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above.
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained.
```
Are the CSI plugins maintained by the community or individual vendors today?

ddebroy replied:
I think it depends on the plugin. For plugins that facilitate interaction with a storage system shipped by a specific storage vendor, e.g. EMC/Netapp/Pure, the CSI plugins are expected to be maintained by that vendor. Same goes for cloud environments like EC2/GCE PD/Azure. For open source storage projects like Ceph, OpenEBS, their CSI plugins will most likely be maintained by an open source community (but may not be the Kubernetes folks).

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, an admission webhook will need to be implemented and deployed that will reject all containers that bind mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have `privileged` flag set in the pod's securityCont ...
```
`Privileged` means a lot more than talking to the proxy on Linux. I wonder if there'd be concern for reusing this filed. 

/cc @tallclair

ddebroy replied:
I wanted to clarify something: in case of Linux, support for privileged containers already exist. So node plugins targeting Linux nodes will not require a proxy for any operations in the context of CSI node plugins. 

The scope of the proxy is thus purely Windows (and maybe expanded to other non Linux OS es later if Kubernetes expands support and those OS es lack privileged support). In case of Windows however, `Privileged` has no relevance today and therefore the proposal to use it.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Test Plan
+
+Unit tests will be added to verify Windows related enhancements in existing Kubernetes components mentioned above.
```
I wish we ran Windows unit tests for PRs

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
```
Any requirement for the new CSI proxy API?

ddebroy replied:
Added.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Upgrade / Downgrade Strategy
+
+If a cluster is downgraded from a version of Kubernetes where the plugin watcher supports Windows nodes to one that does not, existing Windows workloads that were using CSI plugins to access storage will no longer have access to the data. This loss of functionality cannot be handled in an elegant fashion.
```
Could you also add the strategy for upgrading CSI plugins and/or proxy here?

ddebroy replied:
Done

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+The main drawback associated with the approach leveraging csi-proxy.exe is that the life cycle of that binary will need to be managed out-of-band from Kubernetes. However, cluster administrators need to maintain and manage life cycle of other core binaries like kubeproxy.exe, kubelet.exe, dockerd.exe and containerd.exe (in the future). Therefore csi-proxy.exe will be one additional binary that wi ...
+
+The API versioning scheme described above, as of now, will maintain backward compatibility indefinitely. This requires the scope of csi-proxy.exe to be limited to a very scoped down fundamental set of operations. Maintainers therefore will need to be very cautious when accepting suggestions for new APIs and enhancements. This may slow progress at times.
```
Adding another API and maintaining is my biggest concern with the approach. (I think we've also discussed this in the @PatrickLang's doc) Does SIG storage think the API surface will be manageable and stable?

msau42 replied:
What if instead of a common proxy we have a per-driver proxy that is maintained and deployed by each driver vendor?  It pushes the responsibility to the driver maintainers, but API and versioning compatibility is simpler since it's all under the control of one vendor.  However, this does imply we need to give CSI drivers permissions to install binaries directly on the host.

But even for Linux CSI drivers today, some already need to do this to an extent (ie iscsd doesn't work well in a container)

wongma7 replied:
At that point, why not use that same deployment mechanism to install and run the CSI driver directly on the host?

ddebroy replied:
One alternative to having a set of operations as a scoped down API exposed by the proxy is: surfacing a simple `Exec` API in the proxy that accepts arbitrary parameters and returns stdout results. That can supersede all the operations mentioned here as containers will be free to run arbitrary commands on the host. I can list this as an alternative. The main disadvantage is this makes the proxy open to massive abuse.

With this approach + bind mounting a dedicated per-plugin path in the host, plugins can drop arbitrary exes they may need (like `ebsnvme-id.exe` above) and run it through the exec interface if desired. Thoughts?

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Version Skew Strategy
+
+Different nodes in the cluster may be configured with different versions of csi-proxy.exe as part of a rolling upgrade of csi-proxy.exe. In such a scenario, it is recommended that csi-proxy.exe is upgraded first and once that is complete, the CSI Node Plugins that can take advantage of the new version of csi-proxy.exe is rolled out.
```
How would the failure caused by the version skew be manifested to the users?

ddebroy replied:
Will call out explicitly that plugins need to check version of the proxy using the version API call and fail initialization the plugin detects an incompatibility.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Package CSI Node Plugins as containers and deployed as processes running on the host:
+
+With this approach, the container run time is enhanced to be able to launch binaries directly on the host after pulling down a container image and extracting the binary from the image. While usual Kubernetes constructs may be used to launch pods with a special RuntimeClass that can handle launching of the binaries as processes on hosts, Kubernetes won't be able to fully monitor and control the wh ...
```
I think managing lifecycles and logs is possible, but more require work in the container runtime.

The cons mentioned here (i.e., lifecycle management and logs) also present in the approach above "Deploying CSI Node Plugins as binaries and deployed as processes running on the host".

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+### Support for Privileged Operations and Bi-directional mount propagation in Windows containers:
+
+At some point in the future, it is expected that a Windows LTS release will implement support for execution of privileged operations and bi-directional mount propagation from inside containers. At that point, the requirement of a proxy to handle privileged operations on behalf of containers will disappear.
```
I don't think this is confirmed yet (or maybe I've missed the update). I think it'd be good to point out that this option won't be available in a year, hence the need for a interim or even permanent solution.

yujuhong commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+## Infrastructure Needed
+
+The code for csi-proxy as well as the GRPC API will be maintained in a dedicated repo: github.com/kubernetes-csi/windows-csi-proxy 
```
Will there be a csi proxy for linux as well?

ddebroy replied:
At a high level, a CSI proxy for Linux in the context of operations enumerated in the API above, is unnecessary as the operations can be executed from privileged containers directly in Linux. However, as @msau42 pointed out above there are some special scenarios in Linux where a host footprint may be necessary for Linux plugins as well. iSCSI is an interesting corner case today for Linux where the `iscsiadm` binary is typically mounted into the plugin container from the host while interactions with the iSCSI kernel modules is facilitated through `iscsid` running on the host. 

I will call out that the proxy called out in the KEP focusses on Windows today and may be expanded for Linux scenarios in the future. Does that sound good @msau42 @yujuhong?

ddebroy commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+
+CSI Node Plugins listen on domain sockets and respond to CSI API requests sent over GRPC from a container orchestrator like Kubernetes. They are responsible all storage management operations scoped around a specific node that are typically necessary after a CSI Controller Plugin has finished provisioning a Persistent Volume and attached it to the node. In Kubernetes, the CSI Node API calls are in ...
+
+Support for domain sockets has been introduced in Windows Server 2019 and works across containers as well as host and container as long as the processes running in containers are listening on the socket (which is what is necessary in the CSI containerized plugin scenario). Golang supports domain socket operations for Windows since go version 1.12. Thus, some of the fundamental support in the OS a ...
```
Yes, as mentioned above, when the socket is used between a process in the host OS and a process inside a container, the communication works as long as "the process running in containers are listening on the socket". Will try to call it our more clearly.

ddebroy commented:
@msau42 @PatrickLang @yujuhong Thanks for the detailed review so far. I have addressed all the comments thus far and also added a diagram clarifying the interactions on the node side with csi-proxy. PTAL.

ddebroy commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,401 @@
+1. Enable Windows nodes to support CSI plugins so they can surface the above mentioned benefits of CSI plugins to Windows workloads that have persistent storage requirements.
+
+2. The CSI migration initiative (planned to reach beta state in v1.16) aims to deprecate the code associated with several in-tree storage plugins and pave the path for the ultimate removal of that code from Kubernetes core in favor of CSI plugins that implement the same functionality. Windows workloads need to be aligned with the CSI migration effort and cannot depend on environment specific in-t ...
+
```
Done.

ddebroy commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+
+#### Enhancements in CSI Node Plugins
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above. Thus the codebase for most plugins that wishes to support Windows node ...
```
My very rough approximation is more than 20 at this point (targeting Linux). Pretty much all the major storage vendors are working on beta versions of CSI plugins. Some of them are working on multiple plugins targeting different product lines. The major clouds are also developing a CSI plugin and sometimes multiple ones for different storage backends (shared storage vs block).

msau42 replied:
There are 40+ CSI plugins listed publicly: https://kubernetes-csi.github.io/docs/drivers.html

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+1. Ability to handle Windows file paths in the Kubelet plugin watcher for domain sockets on Windows nodes.
+2. Refactor code in the CSI Node Driver Registrar so that it can be compiled for Windows.
+3. Build official CSI Node Driver Registrar container images based on Windows base images.
+4. Refactor code in existing CSI Node Plugins to support Windows. All privileged operations will need to be driven through an API exposed by a "privileged proxy" binary described below.
```
should we leave this step to individual plugin owners? Maybe producing a guide for driver developers to follow is more scalable

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+5. Build CSI Node Plugin container images based on Windows base images.
+6. Create DaemonSet YAMLs referring to CSI Node Driver Registrar container images [3] and CSI Node Plugin container images [5] targeting Windows.
+
+Besides the above enhancements, a new "privileged proxy" binary, named csi-proxy.exe is a key aspect of this KEP. csi-proxy.exe will run as a native Windows process on the Windows nodes configured as a Windows Service. csi-proxy.exe will expose an API (through GRPC over a named pipe) for executing privileged storage related operations on Windows hosts on behalf of Windows containers like CSI Node ...
```
In general can you line break around 80 chars for review ease (can comment on specific lines)

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+
+![Kubernetes CSI Components](https://raw.githubusercontent.com/kubernetes/community/master/contributors/design-proposals/storage/container-storage-interface_diagram1.png?raw=true "Kubernetes CSI Components")
+
+Support for Unix Domain Sockets has been introduced in Windows Server 2019 and works across containers as well as host and container as long as the processes running in containers are listening on the socket. If a process from within a container wishes to connect to a domain socket that a process on the host OS is listening on, Windows returns a permission error. This scenario however does not ar ...
```
What about the proxy? Host listening and the process in the container connecting. Also is there documentation or reasoning on why this restriction exists?

ddebroy replied:
The proxy exposes a Windows named pipe. Will call it out here and mention it is detailed in later sections. Regarding why this restriction exists, given the very specific permission error when trying to connect to a host socket from a containerized process, my assumption is it's a conscious security choice at the OS layer. Unfortunately, this behavior or error is not clearly documented in MSDN.

ddebroy replied:
@liggitt Windows named pipes are an established way to have a container communicate with a host process for a while now in Windows. Support was added back in 2016: https://github.com/moby/moby/pull/19911 in Docker engine and was specifically intended to provide a mechanism similar to Unix Domain Sockets in Windows before support for Domain Sockets was enabled in Windows later in 2019. /cc @PatrickLang for any additional comments.

PatrickLang replied:
Honestly I'm not sure why it's being blocked. It may be a regression in Windows and it's something we can root cause as we implement this. Named pipes and unix domain sockets are functionally equivalent and both work as container mounts, so I think we'll decide on which to enable as part of this work. We've already shown that named pipes work, so if unix domain sockets can be "fixed" we can choose to enable that later.

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+
+Based on the above, we can conclude that some of the fundamental support in the OS and compiler with regards to domain sockets in the context of CSI plugin discovery and a channel for API invocation is present in a stable state in Windows Server 2019 today. Although there are some observed limitations with respect to domain sockets in Windows Server 2019, they are not major blockers in the contex ...
+
+CSI Node Plugins need to execute certain privileged operations at the host level as well as propagate mount points in response to the CSI API calls. Such operations involve: scanning disk identities to map the node OS's view of a disk device to a CSI volume provisioned and attached by CSI controller plugins, partitioning a disk and formatting it when necessary, bind-mounting volumes from the host ...
```
> These operations cannot be invoked from a container in Windows today.

Which of these operations can't be performed from inside a container... all of them?

ddebroy replied:
Correct - all of them. No operations in the host namespace is permitted from inside a container.

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+Golang supports domain socket operations for Windows since go version 1.12. It was found that in Windows, `os.ModeSocket` is not set on the `os.FileMode` associated with domain socket files in Windows. Therefore determining whether a file is a domain socket file using `os.ModeSocket` does not work on Windows.
+
+Based on the above, we can conclude that some of the fundamental support in the OS and compiler with regards to domain sockets in the context of CSI plugin discovery and a channel for API invocation is present in a stable state in Windows Server 2019 today. Although there are some observed limitations with respect to domain sockets in Windows Server 2019, they are not major blockers in the contex ...
+
```
Did we consider instead of using the proxy to deploy the drivers uncontainerized completely? Might be worth noting somewhere if that was an option considered and why it's not the final design

ddebroy replied:
Will add a note here pointing to the alternative design sections later down.

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+
+#### Beta -> GA Graduation
+
+- All in-tree storage plugins that implements support for Windows can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
```
"all"? think we should clarify here, I think all would encompass things like "emptydir", "hostpath".

ddebroy replied:
Good point. Will make this more specific.

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,432 @@
+
+- All in-tree storage plugins that implements support for Windows can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
+- csi-proxy.exe supports v1 stable version of the CSIProxyService API.
+- Successful usage of csi-proxy.exe with support for v1 version of CSIProxyService API in Windows nodes by at least two storage vendors.
```
Not sure if you want to bundle graduation criteria with exact implementation details (maybe implementation might change?), instead it might be helpful to focus on what the end result would be and the ways we can validate those results? Not sure about this comment though, maybe we do want to speak about exact implementation details for more accuracy

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
If a driver is compromised, would exec even pose an even greater risk than the api surface in the main design? Looks like you could expose arbitrary directories to arbitrary places effectively giving you control over the entire system (maybe root through some finite number of steps). Not a security or windows expert though :)

ddebroy replied:
Agree that with a compromised plugin, a lot of damage can be done through the privileged interface. I will clarify that the security concern should not be a reason to pre-clue an `Exec` interface if we do need one.

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
ah this answers one of my previous comments. Might be worth listing the "titles" of the alternate designs at least before getting into the main design so readers know that they're there and where to find them

davidz627 commented on keps/sig-windows/20190714-windows-csi-support.md:
are we aware of any active development in this direction?

ddebroy replied:
There is no publicly declared timeline/commitment around this.

ddebroy commented:
/assign @liggitt

PatrickLang commented:
/lgtm - let's get this draft merged and work on promoting to `implementable` in a new PR

msau42 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskLocations returns locations <Bus, Target, LUN ID> of all disk devices enumerated by Windows
+    rpc GetDiskLocations(GetDiskLocationsRequest) returns (GetDiskLocationsResponse) {}
```
nit: `List` instead of `Get`

msau42 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+
+### Upgrade / Downgrade Strategy
+
+In order to install a CSI Node Plugin or upgrade to a version of a CSI Node Plugin that uses an updated version of the CSIProxyService API not supported by the currently deployed version of csi-proxy.exe in the cluster, csi-proxy.exe will need to be upgraded first on all nodes of the cluster before deploying or upgrading the CSI Node Plugin. In case there is a very old CSI Node Plugin in the clus ...
```
Another strategy for the old apiversion scenario is to have csiproxy support multiple api versions for a few releases according to deprecation policy?

ddebroy replied:
I will call out a deprecation policy mirroring Kubernetes policy but again emphasize that goal should be to maintain compatibility as much as possible.

msau42 commented:
For provisional, this lgtm from storage perspective.

To move to implementable, we should: 
* get approval from the security team on our security strategy
* get approval from sig node on the proxy binary
* anything else?

msau42 commented:
Can you squash your commits?

ddebroy commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+
+Registration of CSI Node Plugins on a Kubernetes node is handled by the Kubelet plugin watcher using the fsnotify package. This component needs to convert paths detected by fsnotify to Windows paths in handleCreateEvent() and handleDeleteEvent() before the paths are passed to AddOrUpdatePlugin() RemovePlugin() routines in desiredStateOfTheWorld. A new utility function, NormalizeWindowsPath(), wil ...
+
+Given `os.ModeSocket` is not set on a socket file's `os.FileMode` in Windows, a specific check for `os.ModeSocket` in handleCreateEvent() will need to be relaxed.
```
It does appear to be a bug in go and I will report it.

ddebroy replied:
Will mention https://github.com/golang/go/issues/33357 in the KEP.

ddebroy commented:
commits squashed

PatrickLang commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+        - name: csi-proxy-pipe
+          hostPath: 
+            path: \\.\pipe\csi-proxy
+            type: null
```
there's no filesystem - do you have another suggestion?

liggitt replied:
absent or "" matches what would be specified in the API (and are equivalent, I think)

msau42 commented:
/lgtm

jingxu97 commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+            - name: plugin-dir
+              mountPath: C:\csi
+            - name: csi-proxy-pipe
+              mountPath: \\.\pipe\csi-proxy
```
here it is using relative path?

ddebroy replied:
@jingxu97 No. Named pipe paths in Windows are always absolute. The format is `\\.\pipe\pipe-name`.

tallclair commented on keps/sig-windows/20190714-windows-csi-support.md:
```diff
@@ -0,0 +1,462 @@
+Support will need to be implemented in AllowsHostVolumePath to handle Windows pipe paths.
+
+#### Mitigation using a webhook
+An admission webhook can be implemented and deployed in clusters with Windows nodes that will reject all containers that mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have privileged flag set in the pod's securityContext specification. This allows the privileged setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide ...
```
I'm not sure how symlinks are handled on windows, but on linux this would be easy to circumvent.

In general, mounting of hostpaths should be thought of as a highly privileged operation, especially if they aren't read-only (can these be mounted read-only?). If the only way to get access to the socket is via a hostpath mount, I'm not too worried.

If we did want to add more enforcement, I'd prefer to add authentication & authorization to the socket itself. E.g. ues token review + subject access review to ensure the container using the proxy is authorized to do so. I'm not sure this is necessary though.

ddebroy replied:
symlinks to named pipes are not supported in Windows.
